hpccm/building_blocks/scif.py

```diff
@@ -0,0 +1,255 @@
+
+        if self.__apptest:
+            apptest = self.__apptest[0].merge(self.__apptest,
+                                              _app=self.__name, _test=True)
+            recipe.append(apptest)
+
+        return recipe
+
+    def __primitives(self, item):
+        """Item is a building block or a primitive.  A building block consists
+        of one or more other building blocks or primitives.
+        Ultimately, every building block consists of primitives.
+
+        "Flatten" the item to a list of its constituent primitives.
+
+        """
+        return [i for i in self.__iter_flatten(item)]
+
+    def __iter_flatten(self, iterable):
+        """Recursively flatten"""
+        try:
+            for i in iter(iterable):
+                for f in self.__iter_flatten(i):
+                    yield f
+        except:
+            # not iterable
+            yield iterable
+
+    def __str__(self):
+        """String representation of the building block"""
+
+        scif_recipe = self.__scif_recipe()
+
+        if (self.__native and
+            hpccm.config.g_ctype == container_type.SINGULARITY):
+            # Take advantage of Singularity's native support for SCI-F.
+            return '\n'.join(str(x) for x in scif_recipe)
+        else:
+            # Generate an external SCI-F recipe file and manually call scif
+
+            # Temporarily switch container format to Singularity to write
+            # the SCI-F recipe file
+            preserved_ctype = hpccm.config.g_ctype
+            hpccm.config.set_container_format('singularity')
+
+            logging.info('Writing {}'.format(self.__scif_file))
+            with open(self.__scif_file, 'w') as f:
+                f.write('\n\n'.join(str(x) for x in scif_recipe))
+
+            # Restore original container format
+            hpccm.config.g_ctype = preserved_ctype
+
+            # Container instructions to copy the SCI-F recipe file
+            # into the container and then run scif
+            c_scif_file = os.path.join('/scif/recipes',
+                                       os.path.basename(self.__scif_file))
+            instructions = []
+            instructions.append(comment('SCI-F "{}"'.format(self.__name)))
+            instructions.append(
+                copy(src=self.__scif_file, dest=c_scif_file))
+            instructions.append(
+                shell(chdir=False,
+                      commands=['scif install {}'.format(c_scif_file)]))
+
+            return '\n'.join(str(x) for x in instructions)
+
+    def runtime(self, _from='0'):
+        """Generate the set of instructions to install the runtime specific
+        components from a build in a previous stage.
+
+        # Examples
+        ```python
+        s = scif(...)
+        Stage0 += s
+        Stage1 += s.runtime()
+        ```
+        """
+
+        if not scif.__runtime_called:
+            scif.__runtime_called = True
+            return str(copy(_from=_from, src='/scif', dest='/scif'))
```

Comment:
This does copy all scifs, which was the intent.

Copying `/scif/apps/[appname]` is an alternative.  I was concerned that this might miss important information in `/scif/data` and/or `/scif/recipes`.  If the scif directory layout were to ever change (unlikely?), this could be fragile.

What is your preference?

Reply:
Just leave it as it is. If this becomes an issue in the future we can tackle it then...